drivers/sd/bcm2711/brcme88c/driver.cpp

```diff
@@ -0,0 +1,452 @@
+static
+void CODE_SEG_PAGE
+DriverExitNormal()
+{
+    PAGED_CODE();
+
+    // This is the code that should run for normal driver exit but
+    // NOT for crashdump-mode driver exit.
+
+    if (g_rpiqFileObject != nullptr)
+    {
+        ObDereferenceObject(g_rpiqFileObject);
+        g_rpiqFileObject = nullptr;
+    }
+
+    TraceLoggingUnregister(g_trace);
+}
+
+_Use_decl_annotations_ static // = PASSIVE_LEVEL
+void CODE_SEG_PAGE
+DriverCleanupNormal(
+    SD_MINIPORT* miniport)
+{
+    PAGED_CODE();
+
+    DriverCleanupCrashdump(miniport);
+
+    LogInfo("DriverCleanup");
+    DriverExitNormal();
+}
+
+_Use_decl_annotations_ static // = PASSIVE_LEVEL
+void CODE_SEG_PAGE
+DriverCleanupCrashdump(
+    SD_MINIPORT* miniport)
+{
+    PAGED_CODE();
+
+    for (UCHAR i = 0; i != miniport->SlotCount; i++)
+    {
+        auto pSlot = miniport->SlotExtensionList[i];
+        if (pSlot != NULL)
+        {
+            auto pSlotExtension = reinterpret_cast<SlotExtension*>(pSlot->PrivateExtension);
+            pSlotExtension->SlotCleanup();
+        }
+    }
+}
+
+_Use_decl_annotations_ extern "C" // = PASSIVE_LEVEL
+NTSTATUS CODE_SEG_INIT
+DriverEntry(
+    DRIVER_OBJECT* driverObject,
+    UNICODE_STRING* registryPath)
+{
+    PAGED_CODE();
+
+    NTSTATUS status;
+
+    // If DriverEntry is called with higher IRQL, assume crash-dump mode.
+    bool const crashDumpMode = KeGetCurrentIrql() != PASSIVE_LEVEL;
+
+    // Logging and regulator control are unavailable in crash-dump mode.
+    if (!crashDumpMode)
+    {
+        // Activate ETW tracing.
+        TraceLoggingRegister(g_trace);
+
+        // BRCME88C: Try to access the RPIQ driver so we can control regulator voltage.
+        ULONG const RpiqDesiredAccess = GENERIC_READ | GENERIC_WRITE;
+        DECLARE_CONST_UNICODE_STRING(rpiqDeviceName, RPIQ_SYMBOLIC_NAME);
+        OBJECT_ATTRIBUTES attributes;
+        InitializeObjectAttributes(
+            &attributes,
+            const_cast<UNICODE_STRING*>(&rpiqDeviceName),
+            OBJ_KERNEL_HANDLE,
+            NULL,
+            NULL);
+        HANDLE rpiqHandle = nullptr;
+        IO_STATUS_BLOCK statusBlock;
+        status = ZwCreateFile(
+            &rpiqHandle,
+            RpiqDesiredAccess,
+            &attributes,
+            &statusBlock,
+            nullptr,
+            FILE_ATTRIBUTE_NORMAL,
+            FILE_SHARE_READ | FILE_SHARE_WRITE,
+            FILE_OPEN,
+            FILE_NON_DIRECTORY_FILE,
+            nullptr,
+            0);
+        if (!NT_SUCCESS(status))
+        {
+            LogWarning("DriverEntry-RPIQ-CreateFileError", TraceLoggingNTStatus(status));
+        }
+        else
+        {
+            status = ObReferenceObjectByHandle(
+                rpiqHandle,
+                RpiqDesiredAccess,
+                *IoFileObjectType,
+                KernelMode,
+                (void**)&g_rpiqFileObject,
+                nullptr);
+            ZwClose(rpiqHandle);
+            if (!NT_SUCCESS(status))
```

Comment:
Consider adding a LogWarning at this point if RPIQ isn't accessible.

Reply:
I'm not sure I understand. I'm already logging a warning on every path where RPIQ is not accessible.

Unless you mean logging in the crashdump path. I could add a LogWarning for that case, but it would be a no-op because logging is always a no-op in the crashdump path (you can't start a new ETW provider when the system is running in crashdump mode).

Reply:
I was concerned about the case where `g_rpiqFileObject` came back as a `nullptr` from `ObReferenceObjectByHandle`. That being said (and now that I have fresh eyes), I suppose it is probably safe to assume `ObReferenceObjectByHandle` returns a valid `g_rpiqFileObject` pointer if status is success. At the time, I was just noticing that future RPIQ accesses were basically gated behind the result of this value so I wanted to be sure a log message was generated at the time of assignment.

Re-reading my comment and the surrounding code, I did a terrible job indicating this. And ultimately it may already be covered by your other warnings here. Sorry about the confusion!